torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -1615,6 +1615,37 @@ bool LoopNest::hasLoopCarriedDependence(For* loop) {
   return false;
 }
 
+bool LoopNest::unsafe_fuseLoops(const std::vector<For*>& loops, For** fused) {
```

Comment:
I'm a bit skeptical about allowing fusion like this when it could be completely wrong to fuse. I understand this API is called `unsafe_` but still..

In addition to not checking for loop-carried dependence, this API does not check if the loops have the same parents, that the loops do not have any statements between them, that the loop bounds are the same. In all these cases, it is just wrong to fuse the loops.

What is the need for this API? Are you trying to have this just to cover the corner cases that the fusion API does not allow?

Reply:
I think it's useful to have both "safe" and "unsafe" versions. In many cases the user might know that fusing two loops is legal where as it could be extremely hard to automatically prove that in compiler. When, say, we're constructing a loop nest for matmul or conv we pretty much might know what can be fused and what can't. The safe version is still important too and can be used in automatic transformations.

Reply:
The only case we cannot prove that fusing is legal is when we report some loop-carried dependences which are false positives. I agree that, in those cases, we can allow "unsafe" fusion without checking for the loop-carried dependence.

We should definitely not allow fusing loops when other conditions are not satisfied, like loops having the same bounds, etc, even in the "unsafe" case.

Reply:
Yeah, that makes total sense.

Reply:
In the "unsafe" version, I tried to remove as much checks as possible and leave the correctness issue to the user.

> We should definitely not allow fusing loops when other conditions are not satisfied, like loops having the same bounds, etc, even in the "unsafe" case.

 Consider symbolic expression bounds, our analysis is not capable to prove the equivalence of such bounds yet. For example, `2s/(s+1)` as loop1 bound and `1` as loop2 bound, `s` is a positive integer - our check does not deem the two equivalent yet.  We can extend our check to cover more cases but should unblock the user before that.

>  this API does not check if the loops have the same parents, that the loops do not have any statements between them

In my understanding, it is not necessary for the loops to have parents to fuse them, also, not necessary to not have any statements between them. Am I missing something?

Reply:
Ah OK, I think I understand the same parent condition. Will add a check for that. But still, the loops can have statements between them as long as there are no data dependence between these statements and the loops.

Reply:
How about we disallow fusion if there stmts between loops (even unsafe version), but provide api to arbitrarily shuffle stmts within a block? It might be less convenient to use since it's broken down into more primitive steps, but it can lead to more composable api in the end.

Reply:
This sounds reasonable. Will make the changes!